This repository contains the associated code for a manuscript under review at Nature Communications. This readme file highlights the function of each script in the repository. 
1. Appetite_data.R - Analysizes and plots appetite data to show differences between geothermal and ambient ecotypes
2. Fst_inversion.R - Calculates Fst within the inversion site
3. Gene_annotation.R - Annotates the Fst data with gene names across the genome for SNPs that are within gene regions. 
4. Iceland_map.R - Creates a map of Iceland
5. MapK_pathway_expression.R - Quantifies gene expression of MAPK associated genes between geothermal and ambient ecotypes across two rearing temperatures
6. Pairise_Fst.R - Calculates pairwise Fst across all populations/ecotypes of Icelandic stickleback
7. Phylo_tree_GBS.R - Creates and vizualises the phylogenetic tree based on GBS data
8. Selective_sweep.R - Calculates selective sweeps
9. afvaper.R - Quantifies regions of genomic parallelism across ecotype pairs
10. cand_gene_function.R - Identifies coding vs regulatory regions for each of the geothermal/ambient ecotype pairs
11. lipid_content.R - Quantifies amounts of body fat between geothermal and ambient ecotypes
12. lostruct_stickleback.R - Runs lostruct to identify putative inversions
13. parallelism.R - Assesses genomic parallelism (multiple ways)
14. rehh_selective_sweep.R - Calcualtes the rehh statistic to assess selection on chromosome XXI
15. Allele_div.R - Produces a PCA for the putative structural variant on chromosome XXI
16. lostruct_functions.R - Contains homemade functions to run the lostruct package more efficiently
17. stickleback_data_explore.R - Big set of code for exploratory data analysis of the WGS data set. Performs PCA and LFMM to assess population structure. 
